### PR TITLE
CIRC-732 Apply side effects to provide a JSON HTTP Response

### PR DESCRIPTION
**Pull request by marcjohnson-kint**
## Purpose
The goal of this work is to start to separate the internal domain of circulation and it's processes from the HTTP API that exposes it externally.

This is the first of a series of pull requests to start to separate the flow of control that the Result class offers from how HTTP responses are done.

A side effect of this work is that the result interface might be simplified and there is a reduction in the workarounds that need to happen to map failures to HTTP.

## Approach
* Started by examining the mapping of success and failure during the check in API
* Extracted a class for representing a HTTP response with a JSON body
* Reworked the existing logic to use that response class
* Use a side effect (rather than mapping magic) to write the HTTP response to vert.x
* Hide the vert.x HTTP Response inside the web context
* Make HttpFailures behave like general HttpResponses
* Extended this approach to other APIs that respond with 200 OK and JSON

#### TODOS and Open Questions
* Extend this approach to other HTTP responses
* Eventually move the HTTP mapping out of the different failures

## Learning
* Trying to hide the failure mapping led to repeated, yet specialised code for many of the APIs

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [x] Were any API paths or methods changed, added or removed?
  - [x] Were there any schema changes?
  - [x] Did any of the interface versions change?
  - [x] Were permissions changed, added, or removed?
  - [x] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/circulation/resources/CheckInByBarcodeResource.java [hunk 1]
```diff
@@ -83,7 +83,8 @@ private void checkin(RoutingContext routingContext) {
       .thenComposeAsync(r -> r.after(
         records -> processAdapter.createOverdueFineIfNecessary(records, context)))
       .thenApply(r -> r.next(requestScheduledNoticeService::rescheduleRequestNotices))
-      .thenApply(CheckInByBarcodeResponse::from)
-      .thenAccept(r -> r.writeTo(routingContext.response()));
+      .thenApply(r -> r.map(CheckInByBarcodeResponse::fromRecords))
+      .thenApply(r -> r.map(CheckInByBarcodeResponse::toHttpResponse))
+      .thenAccept(r -> r.applySideEffect(context::write, context::write));
```
Comment (bohdan-suprun) [inline src/main/java/org/folio/circulation/resources/CheckInByBarcodeResource.java:88]: Does it make sense to move this method somewhere and name it to describe that it is used to write the result?

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/resources/CheckInByBarcodeResource.java:88]: > Does it make sense to move this method somewhere and name it to describe that it is used to write the result?

I'm not sure I follow. Which method are you referring to moving or renaming?

Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/resources/CheckInByBarcodeResource.java:88]: the `r.applySideEffect(context::write, context::write)` method.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/resources/CheckInByBarcodeResource.java]: @bohdan-suprun I've extracted an extra method on the context. What do you think?

I originally named it `writeResponseAsSideEffect`. I felt that was wordy and didn't really add much. Thinking about it now, I could rename it to something like `writeResultToHttpResponse`.

What name would make the most sense to you?

Comment (bohdan-suprun) [reply src/main/java/org/folio/circulation/resources/CheckInByBarcodeResource.java]: The `writeResultToHttpResponse` sounds clear to me.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/resources/CheckInByBarcodeResource.java]: Ok, I'll rename it to that. Thanks

**Pull request merged**